### ISSUE TITLE
QGIS Processing Modeler fails for chained field calculators when using the first results

### ISSUE THREAD

**Issue by PythonicB** [qgis/QGIS#30397]
**Describe the bug**

I am using the processing modeler to create a model to filter values and interpolate them.
I have two pretty similiar calculations i want to chain there. When I do so the second will not run correctly and just output the values from the first calculated attribute. When I run a model for each calculation by itself it works.

**How to Reproduce**

I have points with a value "z". I use a filter to delete spikes. Therefore I have a field calculator with the following code to create h1:

`
IF ($id < 2, h,	
	IF ($id > (count(0) - 1), h, 
		IF( h - 
			( ( attribute( get_feature_by_id(@layer_id, $id - 1), 'h' ) + 
				attribute( get_feature_by_id( @layer_id, $id + 1), 'h' ) ) / 2) > 0.3, 
			( ( attribute( get_feature_by_id(@layer_id, $id - 1), 'h' ) + 
				attribute( get_feature_by_id( @layer_id, $id + 1), 'h') ) / 2 ), 
			h ) 
		) 
	)
`

As a second filter I use the following to create h2:

`IF ($id < 2, h1,	
	IF ($id > (count(0) - 1), h1, 
		IF( h1 - 
			( ( attribute( get_feature_by_id(@layer_id, $id - 1), 'h1' ) + 
				attribute( get_feature_by_id( @layer_id, $id + 1), 'h1' ) ) / 2) > 0.3, 
			( ( attribute( get_feature_by_id(@layer_id, $id - 1), 'h1' ) + 
				attribute( get_feature_by_id( @layer_id, $id + 1), 'h1') ) / 2 ), 
			h1 ) 
		) 
	)
`

h2 will always be h1. I am not sure where the problem is here, but it seems that the results of the first calculation have not been written successfully while the second algorithm is running.

Here is a sample of points I use as input:

`
{
"type": "FeatureCollection",
"name": "points",
"crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
"features": [
{ "type": "Feature", "properties": { "h": 1.0 }, "geometry": { "type": "Point", "coordinates": [ -0.962962962962963, -0.153439153439153 ] } },
{ "type": "Feature", "properties": { "h": 1.2 }, "geometry": { "type": "Point", "coordinates": [ -0.735449735449735, -0.026455026455027 ] } },
{ "type": "Feature", "properties": { "h": 2.0 }, "geometry": { "type": "Point", "coordinates": [ -0.423280423280423, 0.161375661375661 ] } },
{ "type": "Feature", "properties": { "h": 3.0 }, "geometry": { "type": "Point", "coordinates": [ 0.0, 0.407407407407407 ] } },
{ "type": "Feature", "properties": { "h": 1.2 }, "geometry": { "type": "Point", "coordinates": [ 0.407407407407407, 0.677248677248677 ] } },
{ "type": "Feature", "properties": { "h": 1.0 }, "geometry": { "type": "Point", "coordinates": [ 0.775132275132275, 0.896825396825397 ] } }
]
}
`

**QGIS and OS versions**

QGIS version 3.6.3-Noosa
QGIS code revision 0c5774c068
Compiled against Qt 5.11.2
Running against Qt 5.11.2
Compiled against GDAL/OGR 2.4.1
Running against GDAL/OGR 2.4.1
Compiled against GEOS 3.7.2-CAPI-1.11.0
Running against GEOS 3.7.2-CAPI-1.11.0 b55d2125
PostgreSQL Client Verion 9.2.4
SpatiaLite Version 4.3.0
QWT Version 6.1.3
QScintilla2 Version 2.10.8
Compiled against PROJ 5.2.0
Running against PROJ Rel. 5.2.0, September 15th, 2018

Windows 10

**Additional context**

**Comment by nyalldawson** [comment]
The issue here is that expression functions which operate on map layers (including get_feature_by_id ) don't currently have access to any temporary output layers created a step in a model. This applies to a lot of functions, including aggregates.

**Comment by pathmapper** [comment]
Related SE question: 
https://gis.stackexchange.com/questions/379549/qgis-use-of-a-model-output-in-the-same-model-as-a-layer-in-the-field-calculator

**Comment by Jiinglelocks** [comment]
Another related SE question:
https://gis.stackexchange.com/questions/418918/creating-field-with-1-0-values-depending-if-point-sits-inside-polygon-in-qgis

**Comment by Jiinglelocks** [comment]
> The issue here is that expression functions which operate on map layers (including get_feature_by_id ) don't currently have access to any temporary output layers created a step in a model. This applies to a lot of functions, including aggregates.

Can this be countered by making output layers permanent within the model (by filling in the "enter name if this is final result" field)? This missing feature seems like it should be a priority for the graphical modeler. I'd like to try working on it, does anyone know who I can contact that might have some advice about the source code?

**Comment by gioman** [comment]
> does anyone know who I can contact that might have some advice about the source code?

@Jiinglelocks the right place where to discuss such things is the developers mailing list.

**Comment by rebootcat** [comment]
We still can not use algorithm output temporary layer in model

**Comment by kannes** [comment]
> We still can not use algorithm output temporary layer in model

That's correct, you can check by looking at the issues status. As long as it is "open", this feature has not been implemented.

You can fund developer time to implement it if it is important to you.

**Comment by Babelgit** [comment]
It seems the problem is solved now, see these GIS SE answers: https://gis.stackexchange.com/a/469257/88814 and https://gis.stackexchange.com/a/469253/88814
